File: sdk/go/provider/neighbours.go
package provider

import (
	"crypto/rand"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const neighboursBaseURL = "https://neighbours.sh/api/v1"

var neighboursHeaders = map[string]string{
	"Accept":     "application/json",
	"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36",
}

type neighboursDomainListResponse struct {
	Data struct {
		Domains []string `json:"domains"`
	} `json:"data"`
}

type neighboursInboxResponse struct {
	Data []map[string]any `json:"data"`
}

func neighboursRequest(path string, allowNotFound bool) ([]byte, error) {
	req, err := http.NewRequest(http.MethodGet, neighboursBaseURL+path, nil)
	if err != nil {
		return nil, err
	}
	for k, v := range neighboursHeaders {
		req.Header.Set(k, v)
	}

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if allowNotFound && resp.StatusCode == http.StatusNotFound {
		return nil, nil
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("neighbours http %d", resp.StatusCode)
	}
	return raw, nil
}

func neighboursJSON(path string, out any, allowNotFound bool) error {
	raw, err := neighboursRequest(path, allowNotFound)
	if err != nil {
		return err
	}
	if raw == nil {
		return nil
	}
	return json.Unmarshal(raw, out)
}

func neighboursRandomInt(n int) int {
	if n <= 1 {
		return 0
	}
	v, err := rand.Int(rand.Reader, big.NewInt(int64(n)))
	if err != nil {
		return 0
	}
	return int(v.Int64())
}

func neighboursRandomLocal() string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	var b strings.Builder
	b.WriteString("sdk")
	for i := 0; i < 16; i++ {
		b.WriteByte(chars[neighboursRandomInt(len(chars))])
	}
	return b.String()
}

func neighboursAnyString(value any) string {
	switch v := value.(type) {
	case nil:
		return ""
	case string:
		return strings.TrimSpace(v)
	case []any:
		for _, item := range v {
			if hit := neighboursAnyString(item); hit != "" {
				return hit
			}
		}
		return ""
	case map[string]any:
		if s, ok := v["address"].(string); ok && strings.TrimSpace(s) != "" {
			return strings.TrimSpace(s)
		}
		if s, ok := v["text"].(string); ok && strings.Contains(s, "@") {
			return strings.TrimSpace(s)
		}
		if inner, ok := v["value"]; ok {
			return neighboursAnyString(inner)
		}
		return ""
	default:
		return strings.TrimSpace(fmt.Sprint(v))
	}
}

func neighboursFirstAddress(value any) string {
	if hit := neighboursAnyString(value); hit != "" {
		return hit
	}
	return ""
}

func neighboursFlattenMessage(raw map[string]any, recipient string) map[string]any {
	text := neighboursAnyString(raw["text"])
	if text == "" {
		text = neighboursAnyString(raw["snippet"])
	}
	date := neighboursAnyString(raw["date"])
	if date == "" {
		date = neighboursAnyString(raw["received_at"])
	}
	to := neighboursFirstAddress(raw["to"])
	if to == "" {
		to = recipient
	}
	return map[string]any{
		"id":          neighboursAnyString(raw["uid"]),
		"from":        neighboursFirstAddress(raw["from"]),
		"to":          to,
		"subject":     neighboursAnyString(raw["subject"]),
		"text":        text,
		"html":        neighboursAnyString(raw["html"]),
		"date":        date,
		"attachments": raw["attachments"],
	}
}

func neighboursDetail(address, uid string) map[string]any {
	raw, err := neighboursRequest("/inbox/"+url.PathEscape(address)+"/"+url.PathEscape(uid), true)
	if err != nil || raw == nil {
		return nil
	}
	var data struct {
		Data map[string]any `json:"data"`
	}
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil
	}
	if len(data.Data) == 0 {
		return nil
	}
	return data.Data
}

func NeighboursGenerate(domain *string) (*CreatedMailbox, error) {
	var data neighboursDomainListResponse
	if err := neighboursJSON("/config/domains", &data, false); err != nil {
		return nil, err
	}
	domains := make([]string, 0, len(data.Data.Domains))
	for _, item := range data.Data.Domains {
		if hit := strings.TrimSpace(strings.ToLower(item)); hit != "" {
			domains = append(domains, hit)
		}
	}
	if len(domains) == 0 {
		return nil, fmt.Errorf("neighbours: domain list is empty")
	}

	selected := domains[neighboursRandomInt(len(domains))]
	if domain != nil && strings.TrimSpace(*domain) != "" {
		wanted := strings.TrimSpace(strings.ToLower(*domain))
		selected = ""
		for _, item := range domains {
			if item == wanted {
				selected = item
				break
			}
		}
		if selected == "" {
			return nil, fmt.Errorf("neighbours: unsupported domain %s", *domain)
		}
	}

	email := neighboursRandomLocal() + "@" + selected
	return &CreatedMailbox{
		Channel: "neighbours",
		Email:   email,
		Token:   email,
	}, nil
}

func NeighboursGetEmails(email string) ([]NormEmail, error) {
	address := strings.TrimSpace(email)
	if address == "" {
		return nil, fmt.Errorf("neighbours: empty email")
	}

	var data neighboursInboxResponse
	if err := neighboursJSON("/inbox/"+url.PathEscape(address), &data, true); err != nil {
		return nil, err
	}
	if len(data.Data) == 0 {
		return []NormEmail{}, nil
	}

	out := make([]NormEmail, 0, len(data.Data))
	for _, row := range data.Data {
		raw := row
		if uid := neighboursAnyString(row["uid"]); uid != "" {
			if detail := neighboursDetail(address, uid); detail != nil {
				raw = detail
			}
		}
		out = append(out, NormalizeMap(neighboursFlattenMessage(raw, address), address))
	}
	return out, nil
}

File: sdk/npm/src/providers/neighbours.ts
import { randomInt } from 'crypto';
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'neighbours';
const API_BASE = 'https://neighbours.sh/api/v1';

type DomainListResponse = {
  data?: {
    domains?: string[];
  };
};

type InboxAddress = {
  address?: string;
};

type InboxGroup = {
  value?: InboxAddress[];
  text?: string;
  html?: string;
};

type InboxRow = {
  uid?: number | string;
  from?: InboxGroup | InboxAddress[];
  to?: InboxGroup | InboxAddress[] | string[];
  subject?: string;
  text?: string;
  snippet?: string;
  html?: string;
  date?: string;
  attachments?: unknown[];
};

type InboxListResponse = {
  data?: InboxRow[];
};

type InboxDetailResponse = {
  data?: InboxRow;
};

const HEADERS: Record<string, string> = {
  Accept: 'application/json',
  'User-Agent':
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36',
};

function randomLocal(length = 16): string {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let out = 'sdk';
  for (let i = 0; i < length; i++) {
    out += chars[randomInt(chars.length)];
  }
  return out;
}

async function fetchJson<T>(path: string, allowNotFound = false): Promise<T | null> {
  const response = await fetchWithTimeout(`${API_BASE}${path}`, {
    headers: HEADERS,
  });
  if (allowNotFound && response.status === 404) {
    return null;
  }
  if (!response.ok) {
    throw new Error(`neighbours http ${response.status}`);
  }
  return response.json() as Promise<T>;
}

async function listDomains(): Promise<string[]> {
  const data = await fetchJson<DomainListResponse>('/config/domains');
  const domains = (data?.data?.domains || [])
    .map((item) => item.trim().toLowerCase())
    .filter(Boolean);
  if (domains.length === 0) {
    throw new Error('neighbours: domain list is empty');
  }
  return domains;
}

function pickDomain(domains: string[], preferred?: string | null): string {
  if (preferred && preferred.trim()) {
    const wanted = preferred.trim().toLowerCase();
    const hit = domains.find((item) => item === wanted);
    if (!hit) {
      throw new Error(`neighbours: unsupported domain ${preferred}`);
    }
    return hit;
  }
  return domains[randomInt(domains.length)];
}

function firstAddress(value: unknown): string {
  if (!value) return '';
  if (Array.isArray(value)) {
    for (const item of value) {
      const hit = firstAddress(item);
      if (hit) return hit;
    }
    return '';
  }
  if (typeof value === 'string') {
    return value.trim();
  }
  if (typeof value === 'object') {
    const obj = value as Record<string, unknown>;
    if (typeof obj.address === 'string' && obj.address.trim()) {
      return obj.address.trim();
    }
    if (typeof obj.text === 'string' && obj.text.trim() && obj.text.includes('@')) {
      return obj.text.trim();
    }
    return firstAddress(obj.value);
  }
  return '';
}

function flattenMessage(raw: InboxRow, recipient: string): Record<string, unknown> {
  return {
    id: raw.uid == null ? '' : String(raw.uid),
    from: firstAddress(raw.from),
    to: firstAddress(raw.to) || recipient,
    subject: raw.subject || '',
    text: raw.text || raw.snippet || '',
    html: raw.html || '',
    date: raw.date || '',
    isRead: false,
    attachments: Array.isArray(raw.attachments) ? raw.attachments : [],
  };
}

async function fetchDetail(address: string, uid: string): Promise<InboxRow | null> {
  try {
    const data = await fetchJson<InboxDetailResponse>(
      `/inbox/${encodeURIComponent(address)}/${encodeURIComponent(uid)}`,
    );
    if (!data) {
      return null;
    }
    return data.data && typeof data.data === 'object' ? data.data : null;
  } catch {
    return null;
  }
}

export async function generateEmail(domain?: string | null): Promise<InternalEmailInfo> {
  const domains = await listDomains();
  const selectedDomain = pickDomain(domains, domain);
  return {
    channel: CHANNEL,
    email: `${randomLocal()}@${selectedDomain}`,
  };
}

export async function getEmails(email: string): Promise<Email[]> {
  const address = String(email || '').trim();
  if (!address) {
    throw new Error('neighbours: empty email');
  }

  const list = await fetchJson<InboxListResponse>(`/inbox/${encodeURIComponent(address)}`, true);
  if (!list) {
    return [];
  }
  const rows = Array.isArray(list.data) ? list.data : [];
  const emails: Email[] = [];
  for (const row of rows) {
    const uid = String(row.uid ?? '').trim();
    const detail = uid ? await fetchDetail(address, uid) : null;
    emails.push(normalizeEmail(flattenMessage(detail || row, address), address));
  }
  return emails;
}
